cpp/src/arrow/filesystem/filesystem_utils.cc

```diff
@@ -0,0 +1,203 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "arrow/filesystem/filesystem_utils.h"
+#ifdef ARROW_HDFS
+#include "arrow/filesystem/hdfs.h"
+#endif
+#include "arrow/filesystem/localfs.h"
+#include "arrow/filesystem/path_util.h"
+#include "arrow/util/uri.h"
+
+namespace arrow {
+
+namespace fs {
+
+namespace factory {
+
+namespace internal {
+
+class PathInfo {
+ public:
+  PathInfo() {}
+  ~PathInfo() {}
+
+  static Status Make(const std::string& full_path, std::shared_ptr<PathInfo>* path_info) {
+    *path_info = std::make_shared<PathInfo>();
+    RETURN_NOT_OK((*path_info)->Init(full_path));
+    return Status::OK();
+  }
+
+  Status Init(const std::string& full_path) {
+    RETURN_NOT_OK(ParseURI(full_path));
+    full_path_ = std::string(full_path);
+    return Status::OK();
+  }
+
+  FileSystemType GetFileSystemType() { return fs_type_; }
+
+  std::string GetHostName() {
+    auto search = options_.find("host_name");
+    if (search == options_.end()) {
+      return "";
+    }
+    return search->second;
+  }
+
+  int GetHostPort() {
+    auto search = options_.find("host_port");
+    if (search == options_.end()) {
+      return -1;
+    }
+    std::string port_text = search->second;
+    return std::stoi(port_text);
+  }
+
+  std::string GetUser() {
+    auto search = options_.find("user");
+    if (search != options_.end()) {
+      return search->second;
+    }
+    return "";
+  }
+
+  bool GetIfUseHdfs3() {
+    auto search = options_.find("use_hdfs3");
+    if (search != options_.end()) {
+      if (search->second.compare("1") == 0) {
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  int GetRepsNum() {
+    auto search = options_.find("replication");
+    if (search != options_.end()) {
+      if (search->second.empty()) {
+        return 3;
+      } else {
+        return std::stoi(search->second);
+      }
+    } else {
+      return 3;
+    }
+  }
+
+ private:
+  FileSystemType GetFileSystemTypeFromString(const std::string& s) {
+    if (s == "hdfs") {
+      return FileSystemType::HDFS;
+    }
+    if (s == "http") {
+      return FileSystemType::S3;
+    }
+    if (s == "https") {
+      return FileSystemType::S3;
+    }
+    if (s == "file") {
+      return FileSystemType::LOCAL;
+    }
+    if (s.empty()) {
+      return FileSystemType::LOCAL;
+    }
+    return FileSystemType::UNKNOWN;
+  }
+
+  Status ParseURI(const std::string& s) {
+    arrow::internal::Uri uri;
+    RETURN_NOT_OK(uri.Parse(s));
+    fs_type_ = GetFileSystemTypeFromString(uri.scheme());
+    switch (fs_type_) {
+      case FileSystemType::HDFS:
+        options_.emplace("host_name", uri.host());
```

Comment:
I think all the HDFS specifics (options, etc.) should be moved to `hfds.h` / `hdfs.cc`.
For example you could have:
```cpp
class ARROW_EXPORT HadoopFileSystem : public FileSystem {
 public:
  // ...

  /// Create a HdfsFileSystem instance from the given options.
  static Status Make(const HdfsOptions& options, std::shared_ptr<HadoopFileSystem>* out);
  /// Create a HdfsFileSystem instance from the given URI.
  static Status Make(const Uri& uri, std::shared_ptr<HadoopFileSystem>* out);
};
```

(and similarly `LocalFileSystem::FromUri`)

Reply:
I did as that way is because S3Options is configured outside S3FileSystem, so I am thinking to do the same thing here.

Reply:
S3Options is configured in `s3fs.h` / `s3fs.cc`.

Reply:
I am a little confused, if S3Options is configured internally in s3fs.cc why s3fs.h makes both S3Options and Make function which uses S3Options as input ARROW_EXPORT? Doesn’t that means users should create and configure S3Options instance firstly then pass to Make function as parameter? Can you give me a more specific example of how you would like to see ?